granulate_utils/linux/ns.py

```diff
@@ -59,16 +59,20 @@ def resolve_proc_root_links(proc_root: str, ns_path: str) -> str:
     parts = Path(ns_path).parts
 
     path = proc_root
+    seen = set()
     for part in parts[1:]:  # skip the / (or multiple /// as .parts gives them)
         next_path = os.path.join(path, part)
-        if os.path.islink(next_path):
+        while os.path.islink(next_path):
+            if next_path in seen:
+                raise RuntimeError("Symlink loop from %r" % os.path.join(path, part))
+            seen.add(next_path)
             link = os.readlink(next_path)
             if os.path.isabs(link):
                 # absolute - prefix with proc_root
                 next_path = proc_root + link
             else:
                 # relative: just join
-                next_path = os.path.join(path, link)
+                next_path = os.path.join(os.path.dirname(next_path), link)
```

Comment:
I don't undersrand this part, what's the dirname for?

Reply:
`next_path` is the file (or dir) that we're trying to resolve. So if it's a relative link, it's relative to the dir it resides in.
Example: if `/tmp/b` is a link to `a`, then it's a link to `/tmp/a`. If `/tmp/x/b` is a link to `../a` then it's a link to `/tmp/x/../a`.

Reply:
This basically means that relative links never worked, then. Right?

Reply:
They did because we used to join `path` and not `next_path` (and `next_path` was originally just `join(path, part)` so `dirname(next_path) == path`. The reason I had to change it is because we can now resolve multiple times inside the loop and `dirname(next_path)` won't necessarily stay the same as `path` :)

Reply:
I see. Kk